### PR TITLE
CI docs

### LINKED ISSUE #181: Document CI
Both internal and external contributors could benefit from more detailed documentation about our CI. All the public documentation I found was item 15 here:

https://ignitionrobotics.org/docs/all/contributing#contributing-code

Things to document: 

* All the checks we have (Jenkins, GitHub actions, DCO...)
* What are `required` checks
* How to inspect results on Jenkins and GitHub actions
* Retriggering CI (`@osrf-jenkins run tests`, retry / rerun buttons)
* Triggering builds outside of pull requests
* Checking the current status of CI for a development branch

### PR DESCRIPTION
**Pull request by chapulina**
Closes #181

This is mostly a brain dump of all the key information I could think of related to our CI.

Feedback is very appreciated :bowing_woman: 

I'll work on a way of rendering this in the main site, see #176.

**Note to maintainers**: Remember to use **Squash-Merge**

🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸

### CODE REVIEW

File: ci.md
Comment (mabelzhang) [inline ci.md:126]: Should we add a note about how to get to the History page of the test? I found that tip to be very helpful in telling me exactly where to look to determine whether it's a new failure or preexisting. If I remember correctly, there were two or more places that say "History" you can click on, and one of them would give me the graph I want. The page seemed different for different test sometimes, and I don't know why that should be.

File: ci.md
Comment (mabelzhang) [inline ci.md:71]: I don't know if we have information on DCO for beginner contributors, but I was very confused in the beginning, when GitHub gave me a command to run to fix DCO, and it ended up signing alllllll the commits, even ones I merged in from other people. To fix the fix, I basically rechecked out the repo... There are more helpful commands online than the one GitHub gives.

File: ci.md
Comment (mabelzhang) [inline ci.md:23]: Should we define "stable," since it's referred to so many times? Does stable == release?

File: ci.md
Comment (mabelzhang) [inline ci.md:56]: Is test coverage only for Bionic and not Focal?
The Bionic and Focal items are exactly the same other than the distro name. It felt dizzy reading this dense technical text twice. Is it possible to combine them?

File: ci.md
Comment (mabelzhang) [inline ci.md:86]: Could there be a one-sentence description in that issue saying, the checked things are required? That wasn't obvious from the title, with "check" meaning different things in "CI checks" and "All checked" in the checkbox lists. A bit of a mind pause there.

File: ci.md
Comment (mabelzhang) [inline ci.md:93]: ✅ should be ✔️
Just kidding..........

File: ci.md
Comment (mabelzhang) [inline ci.md:180]: I think I'm just not understanding this because I've never had to do this. Is this for if you are adding things that will require new dependencies, and this adds dependencies to CI so that you can test that it works, before officially adding a dependency to the repo?
Is there an example for this bullet? I think the example in the next bullet is helpful (though I've never had to do that either, so can't comment on that).

**Review by mabelzhang — COMMENTED** [review]
This is useful. Thank you for putting it together!
I put in a few comments for things I've seen in the past, but that might not be a good reflection of whether this would make sense to someone who's never done them. I didn't understand some items I've never had to do, maybe just because we learn best by doing.

Overall, it's pretty dense reading, and I see it as a reference guide for experienced contributors, or a big learning curve for a beginner.

File: ci.md
Comment (chapulina) [reply ci.md:71]: We have some guidance in the contribution guide, so I didn't want to repeat that information here. I'll add a link: https://ignitionrobotics.org/docs/all/contributing#process

File: ci.md
Comment (chapulina) [reply ci.md:23]: Good point, I'll link to these docs: https://github.com/ignitionrobotics/docs/blob/master/release.md#type-of-releases

File: ci.md
Comment (chapulina) [reply ci.md:56]: Yeah the only difference is that we only run coverage for one of them, because we can't keep track of multiple coverage results for the same branch. I agree that combining should make it easier to read :+1:

File: ci.md
Comment (chapulina) [reply ci.md:86]: I added a few sentences :sweat_smile: Once this guide is live, we can link to it from that issue for better context.

File: ci.md
Comment (mabelzhang) [reply ci.md:93]: Ohhh interesting. On Ubuntu Chrome, the former is white and latter green. On Android, the former is green and latter black 🤷‍♀️

Comment (chapulina) [reply ci.md:93]: I was annoyed by the same thing :sweat_smile: Gotta ask the Unicode Consortium to be more specific about check colors, those render differently according to the platform :grimacing: 

image

File: ci.md
Comment (chapulina) [reply ci.md:180]: I'll add more information about when this may be needed. It's tough to be concise and thorough at the same time :sweat_smile: 

One possible reason to use this is testing changes across repos. If your `ign-gazebo` PR needs a change from `ign-physics`, you could build `ign-physics` from source to check it works on CI before even opening the `ign-physics` PR, or making a release.

File: ci.md
Comment (adlarkin) [inline ci.md:38]:
```suggestion
    * Library being tested and its Ignition dependencies are compiled using [colcon](https://colcon.readthedocs.io/en/released/#).
```

**Review by nkoenig — APPROVED** [review]
Minor suggestions.

File: ci.md
Comment (adlarkin) [inline ci.md:66]: I think this is the first time we mention `push` checks - this is just a check for a commit, right? Maybe we should mention that here, or in line 49 where we first discuss what commits trigger.

File: ci.md
Comment (adlarkin) [inline ci.md:68]: I find it a little awkward that just before this, we discuss bionic vs focal checks for GitHub Actions, but then here, we are discussing other checks which seem to be run on both bionic and focal. As stated in https://github.com/ignitionrobotics/docs/pull/182/files#r627050201, it seems like the bionic vs focal checks section has some duplicate information, so maybe we can merge that with the comments in this section for an easier read.

File: ci.md
Comment (adlarkin) [inline ci.md:44]: It would be nice to include a summary/table at the end of this section that lists what checks are different between Jenkins and GitHub actions.

File: ci.md
Comment (mabelzhang) [reply ci.md:180]: Ah okay. Then I've actually come across the need for that but wasn't aware how it can be done. I think saying that scenario helps.

File: ci.md
Comment (adlarkin) [inline ci.md:143]: How do external contributors report to the build farmer? Should they create/add to an issue? If so, where can this issue be created/found?

File: ci.md
Comment (adlarkin) [reply ci.md:180]: On a related note, I just looked at this link, and I think it'd be nice to clarify what to do for `Be sure to add the apt dependencies of dependencies build from source to apt-dependencies`, because I honestly have no clue how to do this after reading that sentence :upside_down_face:

File: ci.md
Comment (jennuine) [inline ci.md:97]: If I just started and had no experience with CI, I don't think I would have known where the build queue was

```suggestion
Depending on the library and on the build queue (from our [Jenkins server](https://build.osrfoundation.org/)), checks can be in a pending
```

File: ci.md
Comment (jennuine) [inline ci.md:140]:
```suggestion
    * Actions: When there are code check failures, the build is marked red ❌.
```

File: ci.md
Comment (jennuine) [inline ci.md:143]: Might be good to add a short statement on who the build farmer is and what they do? :farmer:

File: ci.md
Comment (jennuine) [reply ci.md:180]: If we later want to open a PR, would we have to remove the dependency file?

**Review by jennuine — COMMENTED** [review]
:smile: Thanks for documenting! Looks good so far, left some minor comments

**Review by chapulina — COMMENTED** [review]
Thank you for all the fantastic feedback! I've refactored a lot of the docs. Another review pass would be appreciated!

File: ci.md
Comment (chapulina) [reply ci.md]: I turned the entire section into a table :smirk:

File: ci.md
Comment (chapulina) [reply ci.md]: Ok, this _should_ make things clearer: https://github.com/ignition-tooling/action-ignition-ci/pull/37

File: ci.md
Comment (adlarkin) [inline ci.md:59]: This is a nice chart! Should we mention why we are running tests on both GitHub actions `Ubuntu CI` and Jenkins `<library>-ci-pr_any-ubuntu-auto-amd64`? Readers might wonder why Ubuntu tests are being done twice (and to be honest, I've always wondered this myself as well :upside_down_face:)

File: ci.md
Comment (adlarkin) [inline ci.md:146]:
```suggestion
* There are **infrastructure failures**. Contributors should leave a comment to
    maintainers (this can be done by leaving a comment on the PR), who will report these to the build farmer.
```

**Review by adlarkin — APPROVED** [review]
New changes look great, thanks @chapulina!

**Review by jennuine — APPROVED** [review]
Looks great! :tada:

File: ci.md
Comment (chapulina) [reply ci.md:59]: Added in 191a329

File: ci.md
Comment (scpeters) [inline ci.md:208]: I would leave a warning that this sometimes causes problems and link to https://github.com/ignition-tooling/release-tools/issues/242

**Comment by chapulina** [comment]
Thanks, @scpeters . Addressed your comments on 05b5573

**Review by scpeters — APPROVED** [review]

**Comment by scpeters** [comment]
@j-rivero I'm going to merge this now; if you see anything you'd like changed, let's do it in another pull request

**Comment by scpeters** [comment]
> For this to work, these 2 PRs need to go in:
> 
> * https://gitlab.com/ignitionrobotics/web/web-server/-/merge_requests/18
> * https://gitlab.com/ignitionrobotics/web/web/-/merge_requests/31

never mind, I'm waiting on these PRs before merging

**Comment by j-rivero** [comment]
I've pushed some changes, nothing major, clarifications and improvements:
 * 6934df3 we are supporting two types of Windows builds, describing both in the docs
 * d570eb4 bot is not osrf-build but osrfbuild
 * 68af95a extra information to help to locate the relevant options
 * 997803c release page seems not to work yet, used github markdown as woraround

If someone can please give a look to these changes, we are good to merge if they are fine.

File: contributing.md
Comment (scpeters) [inline contributing.md:672]: I'm not sure if this is the preferred link format; we may want to direct to ignitionrobotics.org/docs

**Review by scpeters — COMMENTED** [review]
> I added the `unlisted` field to this page, as well as `release` and `architecture`, so that they're generated, but they're not listed on the main `/docs` page. (closes #176)
> 
> For this to work, these 2 PRs need to go in:
> 
> * https://gitlab.com/ignitionrobotics/web/web-server/-/merge_requests/18
> * https://gitlab.com/ignitionrobotics/web/web/-/merge_requests/31

I believe these web PR's haven't been merged yet. Do we need to wait for those?

File: contributing.md
Comment (j-rivero) [reply contributing.md:672]: > I believe these web PR's haven't been merged yet. Do we need to wait for those?

arg, sorry missed the comment. We can wait for them to be merged yes.

> I'm not sure if this is the preferred link format; we may want to direct to ignitionrobotics.org/docs

found it in a different doc and though it can serve as workaround. I'm going to revert the change and we'll wait for the web-server PRs to be merged.

File: ci.md
Comment (jennuine) [inline ci.md:173]: I thought it was `@osrf-jenkins`? Is there a difference between the two?

**Comment by j-rivero** [comment]
**Mental note:** we need to wait for https://github.com/ignitionrobotics/docs/issues/176 before merging.

**Review by nkoenig — APPROVED** [review]

**Pull request merged**
